Update scripting.md to update details and fix inconsistencies

Basssiiie opened the pull request:
Hey all, 

I've updated the `scripting.md` manual file to add some new information and fix a few inconsistencies.

- Add details about the "Open custom content folder" option in the title screen.
- Update details regarding mutating game state for local plugins (build-in game actions are allowed).
- Added a link about Duktape's performance tips.
- Replaced "mimicking internal data structures" with "provide access to" as nowadays we are more critical about mimicking them in regards to future compatibility.
- Replaced a mentioned ride breakdown hook with a ride ratings hook, as the ride breakdown hook does not exist.
- Updated the "game actions" question to first be about built-in game actions, and go in detail about custom game actions later (instead of vaguely mixing them).
- Replaced `network.createServer` with `network.createListener` as the first one does not exist and the second one has all the members used in the example.
- Clarified some details about multiple files and libraries, as I've seen some struggles with that in the past. Also removed the global scope recommendation, as it only works in specific cases, can be quite problematic, and IMHO should be removed completely in the future.
- Clarified a FAQ about missing ES6+ functions and features in Duktape.
- Clarified more specifically what `minApiVersion` does and added a link.

Would love to hear any feedback on this, and if there's other things that should be clarified/improved.

Thank you for your time!

Basssiiie commented:
@Sadret you might be interested as well.

Gymnasiast commented on distribution/scripting.md:
```diff
@@ -8,14 +8,18 @@ Each script is a single physical javascript file within the `plugin` directory i
 - Mac: `/Users/YourName/Library/Application Support/OpenRCT2`
 - Linux: `$XDG_CONFIG_HOME/OpenRCT2` or in its absence `$HOME/.config/OpenRCT2`
 
+You can also find the user directory via the "Open custom content folder" option under the red toolbox button in the game's title screen.
```
```suggestion
You can also find the user directory via the “Open custom content folder” option under the red toolbox button in the game's title screen.
```

Gymnasiast commented on distribution/scripting.md:
```diff
@@ -8,14 +8,18 @@ Each script is a single physical javascript file within the `plugin` directory i
 * Intransient
 
-Local scripts can **not** alter the game state. This allows each player to enable any local script for their own game without other players needing to also enable the same script. These scripts tend to provide extra tools for productivity, or new windows containing information.
+Local scripts can **not** alter the game state directly. This allows each player to enable any local script for their own game without other players in a multiplayer server also needing to also enable the same script. Local scripts can interact with the game by mimicing a player's actions through the built-in "game actions". These scripts tend to provide extra tools for productivity, or new windo ...
```
```suggestion
Local scripts **cannot** alter the game state directly. This allows each player to enable any local script for their own game without other players in a multiplayer server also needing to also enable the same script. Local scripts **can** interact with the game by mimicking a player's actions through the built-in “game actions”. These scripts tend to provide extra tools for productivity, or new windows containing information.
```

Gymnasiast commented on distribution/scripting.md:
```diff
@@ -81,15 +85,13 @@ Yes, but the performance would be so dire that it would be a waste of their time
+Scripts can consist of any behaviour and have a large memory pool available to them. The speed will vary depending on the hardware and system executing them. The scripts are interpreted, so do not expect anywhere close to the performance of native code. In most scenarios this should be satisfactory, but a random map generator, or genetic algorithm for building roller coasters might struggle. Like ...
 
-The APIs for OpenRCT2 try to mimic the internal data structures as close as possible but we can only add so many at a time. The best way to grow the plug-in system is to add APIs on-demand. So if you find an API is missing, please raise an issue for it on GitHub and also feel free to submit a pull request afterwards.
+The APIs for OpenRCT2 try to provide access to the game's data structures as much as possible but we can only add so many at a time. The best way to grow the plug-in system is to add APIs on-demand. So if you find an API is missing, please raise an issue for it on GitHub and also feel free to discuss it on our Discord and to submit a pull request afterwards.
```
```suggestion
The APIs for OpenRCT2 try to provide access to the game’s data structures as much as possible but we can only add so many at a time. The best way to grow the plug-in system is to add APIs on-demand. So if you find an API is missing, please raise an issue for it on GitHub and also feel free to discuss it on our Discord and to submit a pull request afterwards.
```

Gymnasiast commented on distribution/scripting.md:
````diff
@@ -81,15 +85,13 @@ Yes, but the performance would be so dire that it would be a waste of their time
-In case there are breaking Api changes plugins can use this to keep the old Api behavior, as an example in version 34 ```Entity.Type``` would no longer return "peep" for
-guests and staff, instead it would return either "guest" or "staff", so if your plugin expects "peep" you can specify the version 33 to keep the old behavior. See the list
-of breaking changes. If this is not specified it will default to version 33, it is recommended to specify the current api version.
+In case there are breaking Api changes plugins can use this to keep the old API behavior. For example in version 34 ```Entity.type``` would no longer return "peep" for guests and staff, instead it would return either "guest" or "staff", so if your plugin expects "peep" you can specify the version 33 to keep the old behavior. See the list of breaking changes. If this is not specified it will defau ...
````
```suggestion
In case there are breaking API changes, plugins can use this to keep the old API behaviour. For example in version 34 ```Entity.type``` would no longer return `peep` for guests and staff, instead it would return either `guest` or `staff`, so if your plugin expects `peep` you can specify the version 33 to keep the old behaviour. See the list of breaking changes. If this is not specified it will default to version 33 and the plugin will log an error on startup, it is recommended to specify the current API version.
```

Also: this refer to a list of breaking changes, but does not specify where.

Gymnasiast commented on distribution/scripting.md:
````diff
@@ -113,11 +115,11 @@ context.subscribe('interval.day', function() {
 ```
 
-Other hooks include receiving a chat message in multiplayer, or a ride breaking-down.
+Other hooks include receiving a chat message in multiplayer, or a ride's ratings have been recalculated.
````
```suggestion
Other hooks include receiving a chat message in multiplayer, or recalculation of a ride’s ratings.
```

Gymnasiast commented on distribution/scripting.md:
```diff
@@ -129,6 +131,8 @@ Game actions allow you to define new actions (with a new permission slot) that p
 
 This sequence of actions ensures that every player execute the action exactly in the same way on the exact same game tick. It also allows the server to validate that the action is allowed and does not fail due to permission or another player executing a conflicting action just before it. This is why there is a noticeable delay when constructing in multiplayer games, as the client has to wait for  ...
 
+Besides built-in game actions, remote scripts can also register their own game actions ("custom game actions"). These custom actions allow custom game state mutations to be executed in sync for all players on the server. Beware that the plugin will also be responsible for proper permission checks, as custom game actions have no permission checks by default.
```
```suggestion
Besides built-in game actions, remote scripts can also register their own game actions (“custom game actions”). These custom actions allow custom game state mutations to be executed in sync for all players on the server. Beware that the plugin will also be responsible for proper permission checks, as custom game actions have no permission checks by default.
```

Gymnasiast commented:
Overall, the changes look fine, left a few grammar and typographical remarks, as well as one remark about a seemingly missing(?) list.

Basssiiie commented:
@Gymnasiast Thanks! All good points. I've applied the feedback and added a local link to the breaking changes section (it's near the beginning). 🙂

Sadret commented on distribution/scripting.md:
```diff
@@ -8,14 +8,18 @@ Each script is a single physical javascript file within the `plugin` directory i
 * Intransient
 
-Local scripts can **not** alter the game state. This allows each player to enable any local script for their own game without other players needing to also enable the same script. These scripts tend to provide extra tools for productivity, or new windows containing information.
+Local scripts **cannot** alter the game state directly. This allows each player to enable any local script for their own game without other players in a multiplayer server also needing to also enable the same script. Local scripts **can** interact with the game by mimicking a player's actions through the built-in “game actions”. These scripts tend to provide extra tools for productivity, or new w ...
```
The section "What does the error 'Game state is not mutable in this context' mean?" is similarly wrong.

Basssiiie replied:
Yes, I can clarify the difference between singleplayer and multiplayer as you mentioned, as local vs. remote has no differences there. 

For multiplayer I do think "remote" by default is better. It's better that the plugin you bring doesn't load at all than that it loads but breaks/errors because the plugin developer didn't take mutable context in account. Having it as "remote" by default is more intuitive for users and developers to understand why it doesn't load in multiplayer if multiplayer hasn't been accounted for, as it's "oh I need to do something extra for multiplayer" (as you explained).

Note that remove server plugins can exist without custom game actions; e.g. UI popups/welcome messages when a player joins, spam protection, tools to kick/ban players, block certain built-in game actions by subscribing to the hooks, run certain other code tied to hooks ("interval.day", "ratings.calculate" etc.).

Sadret replied:
I think the main point of the local vs remote question is "What does my plugin want to do?" and not "Should my plugin target singleplayer or multiplayer?" and I think in the past there was a lot of confusion about this, maybe because the naming could suggest that "local" is for single player and "remote" is for multiplayer, which is absolutely not the case.

Sadret commented on distribution/scripting.md:
```diff
@@ -129,6 +132,8 @@ Game actions allow you to define new actions (with a new permission slot) that p
 
 This sequence of actions ensures that every player execute the action exactly in the same way on the exact same game tick. It also allows the server to validate that the action is allowed and does not fail due to permission or another player executing a conflicting action just before it. This is why there is a noticeable delay when constructing in multiplayer games, as the client has to wait for  ...
 
+Besides built-in game actions, remote scripts can also register their own game actions (“custom game actions”). These custom actions allow custom game state mutations to be executed in sync for all players on the server. Beware that the plugin will also be responsible for proper permission checks, as custom game actions have no permission checks by default.
```
Technically, any plugin can define and use custom actions, not just remote plugins.

Basssiiie replied:
Technically yes, but practically there's not much benefit. In multiplayer clients who don't have the local plugin won't execute the custom action.

Basssiiie commented:
I've done the following changes:

- Renamed all cases of "plug-in" to "plugin".
- Added `minApiVersion` to the `registerPlugin()` example. 
- Moved the segment about plugin script types to after the writing plugins part and added some extra headers.
- Clarified the differences in behaviour of plugin script types between singleplayer and multiplayer.
- Moved the game actions FAQ to the main section about writing plugins, as I think it's important enough to have it as its own section.
- Moved the FAQ section about `minApiVersion` to be next to the `targetApiVersion`.
- Added some horizontal lines between each FAQ for readability and separation.

@Sadret does this version sound more fitting? 🙂

Sadret commented on distribution/scripting.md:
```diff
@@ -45,16 +34,74 @@ registerPlugin({
+
+## Game actions
+
+Game actions are actions that players can invoke in games. Here is an example flow of a game action such as opening the park:
```
Please add some intuition how to use actions (especially compared to direct data access):
- for everything that the player can do to change the game state, there is a game action
- plugins can execute any of these actions as if they were a player (it is actually never mentioned here)
- **using actions is the prefered way to change the game state**, since it automatically ensures multiplayer-safety and sanity checks (cash, clearance, ...)
- but, if the plugin really needs to use direct game state changes, it is strongly advised to make the plugin remote and wrap all game state changes in custom actions. Otherwise, the plugin will not work in multiplayer and likely cause disconnects. Even if you as the plugin developer do not want to specifically target multiplayer, it is still better to follow these guidelines to prevent disconnects of your users.

Please copy or modify as you see fit.
As a side note, if it wouldn't absolutely destroy backwards compatibility, I would really like to disable all direct game state changes outside of custom actions.

Sadret reviewed (changes requested):
I like the improvements but left some more comments.

Basssiiie commented on distribution/scripting.md:
```diff
@@ -45,16 +34,74 @@ registerPlugin({
+
+In multiplayer local and remote plugins are different in when they are loaded and how they can alter the game state.
+
+**Local plugins** will load on any client in multiplayer, which will allow each player to bring any local script for their own game without other players in a multiplayer server also needing to also enable the same script.
```
What is the misleading part? Also not sure what "as usual" refers too? Singleplayer? That may be misleading because local plugins are not completely "as usual" as singleplayer.

I'll fix the double "also" though, bad copy paste mistake. 🙂

Sadret replied:
"Local plugins will load on any client in multiplayer" can be interpreted as any client that joins will load your local script.
Yes, with usual I meant compared to singleplayer.

Basssiiie replied:
Good point. Added a few extra words that will hopefully clarify that confusion. 🙂

Basssiiie commented:
@Sadret I didn't write the game actions section, but merely moved it. Either way I've updated/rewrote it with your recommendations. 🙂

Furthermore:

- Moved the hot reload and plugin licence to before the script types section, to have the script types section flow more nicely right into the game actions section.
- Added a FAQ about what is recommended for people who only care about singleplayer.
- Added a few local links to the game actions section.

Sadret reviewed (approved):
Looks good to me now, just found one more grammar thing.

Basssiiie commented on distribution/scripting.md:
```diff
@@ -45,16 +34,89 @@ registerPlugin({
+
+This sequence of steps ensures that every player client executes the game action exactly in the same way on the exact same game tick (synchronized execution). Furthermore, it allows the server to validate that the action is allowed and does not fail due to permission or another player executing a conflicting action just before it. This is why there is a noticeable delay when constructing in multi ...
+
+Plugins can also use game actions to alter the game in the same way players can. Game actions are the **recommended approach** to mutating game state, as it automatically ensures multiplayer-safety and various sanity checks for money, clearance checks, and others.
```
Fixed, thanks! 🙂

Gymnasiast commented on distribution/scripting.md:
```diff
@@ -69,40 +131,83 @@ JavaScript is a very mature and flexible language with a large, if not the large
+    authors: ['Your Name'],
+    type: 'remote',
+    minApiVersion: 7, // OpenRCT2 plugin API v7 or higher is required to run your plugin
+    main: main
```
Plugins should also specify their licence. (We generally recommend `MIT` as the default licence here.)

Gymnasiast commented on distribution/scripting.md:
```diff
@@ -69,40 +131,83 @@ JavaScript is a very mature and flexible language with a large, if not the large
+    name: 'Your Plugin',
+    version: '1.2', // Your plugin version
+    authors: ['Your Name'],
+    type: 'remote',
```
```suggestion
    type: 'remote',
    licence: 'MIT', // Specify your licence here (MIT is just a recommendation). You can use the `SPDX` identifier, or point to a file that contains the licence text.
```

Gymnasiast commented on distribution/scripting.md:
```diff
@@ -69,40 +131,84 @@ JavaScript is a very mature and flexible language with a large, if not the large
+    version: '1.2', // Your plugin version
+    authors: ['Your Name'],
+    type: 'remote',
+    licence: 'MIT', // Specify your licence here (MIT is just a recommendation). You can use the `SPDX` identifier, or point to a file that contains the licence text.
```
```suggestion
    licence: 'MIT',
```
